### ISSUE TITLE
Rethinking the Database <-> Engine <-> Catalog mapping

### ISSUE THREAD

**Issue by shangyian** [DataJunction/dj#382]
I started thinking through the mappings between `Database`, `Engine`, `Catalog` and how they all fit together and drew a quick sketch to solidify things in my head:

<img src="https://user-images.githubusercontent.com/9524628/222763225-6b9d86db-6f12-4933-bc4e-52ca5d992ec8.png" width="550" />

Essentially, there are many different databases, some of which include both data storage + an engine (like Postgres and Druid), others of which only have the data storage component (like the data warehouse). Some of the data stores have concepts of catalogs and schemas (Postgres, data warehouse), but others don't (Druid). 

Perhaps instead of having a `Database` concept, we can have a `DataStore` concept, which more cleanly reflects how some components should be represented in this system. So something like this:

<img src="https://user-images.githubusercontent.com/9524628/222765116-98d22bef-8bbe-499f-9f70-33cd134d43ff.png" width="400" />

* A **DataStore** (example values `druid`, `postgres`, `warehouse`) has one or more catalogs associated with it. 
* A **Catalog** has one or more engines (example values `druid`, `postgres`, `spark`, `trino`) associated with it.

What do you folks think?

**Comment by samredai** [comment]
I think this is great. Thanks for the chat earlier and clearing up my funky understanding of postgres 😂 . Adding the datastore ultimately makes sense to me as the final design, but I think it should be paired with adding support for multi-catalog queries. So maybe we can go with the `Catalog` + `Engine` design for now, and then whenever we're ready to add muti-catalog support we tackle adding the Datastore level.

**Comment by shangyian** [comment]
Definitely! This doesn’t need to be done for the alpha release, just wanted to clear up these concepts first. 

@betodealmeida it’d be great to hear your thoughts as well!

**Comment by betodealmeida** [comment]
This is super interesting... I always like to think about how to apply this to more unconventional databases:

- SQLite: should the catalog be a file (eg, `database.sqlite`)? And the data store be a filesystem?
- Google Sheets: there's no catalog nor schema, so should have `main` for both?
- Druid: similarly, there's no catalog.. should we use `main`? Or `default`?

**Comment by samredai** [comment]
I opened PR [#8](https://github.com/DataJunction/djqs/pull/8) against the DJQS repo to make the examples more clearly match this design. `djdb` is the postgres database defined as a DJQS catalog. The `uri` is defined as an attribute on the `postgres` engine (that's then attached to the `djdb` DJQS catalog). It doesn't yet include any concept of a datastore which we can add when we cross that bridge...
